### ISSUE TITLE
Incorrect Scope Format in API Response: Single String Instead of Array for Multiple Scopes

### ISSUE THREAD

**Issue by vamsii777** [vamsii777/vapor-oauth#11]
There appears to be a discrepancy in how scopes are being handled within the Scope Validator of the vapor-oauth project. While the `clientScopes` are correctly retrieved as an array (e.g., `["admin", "openid"]`), the scopes fetched via the API are concatenated into a single string (`"admin,openid"`).

### Specific Problem
The core issue arises from the way scopes are being processed and returned. The expected behavior is to have the scopes returned as an array of strings, which facilitates accurate scope comparisons. However, currently, the API merges all scopes into a single string, thus causing comparison mismatches and functional limitations.

### Examples and Implications
- **Example 1:** As demonstrated in the first screenshot, `clientScopes` retrieves `["admin","openid"]`, which is the expected format.
- **Example 2:** The API response, as shown in the second screenshot, provides the scopes as `"admin,openid"`. This format is technically correct but not suitable for direct comparisons.
- **Failure Case:** This issue is exemplified in the third screenshot, where an attempted comparison of individual scopes (e.g., `"openid"` against `"openid, admin"`) fails due to the string format.

The consequence of this issue is significant: it restricts users to request only one scope at a time, limiting the functionality and flexibility of scope management within the application.

### Suggested Fix
The API should be modified to split the scopes string into an array before any comparison or further processing. This would align the format with the expected array structure and ensure accurate scope comparisons and validations.

### Additional Context
This issue was first identified in a discussion thread (refer to the original post by @mynona in https://github.com/vamsii777/vapor-oauth/discussions/5#discussioncomment-7938901). The provided screenshots and descriptions therein offer further insights into the problem and its implications.

**Comment by mynona** [comment]
Hi @vamsii777, 

I checked the scope issue in more detail and found that there is no issue.

AuthorizeGetHandler.swift

```
private func validateRequest(_ request: Request) async throws -> (Response?, AuthorizationGetRequestObject?) 

(...)

let scopes: [String]
        
        if let scopeQuery: String = request.query[OAuthRequestParameters.scope] {
            scopes = scopeQuery.components(separatedBy: " ")
        } else {
            scopes = []
        }
(...)

```

As you can see it accepts scopes sent from the client only separated by spaces (" "). 

I found different specifications as we use ["scope","scope"]; "scope,scope"; "scope scope" as possibilities. According to
https://www.rfc-editor.org/rfc/rfc6749#section-3.3 
it says:

> The value of the scope parameter is expressed as a list of space-
>    delimited, case-sensitive strings.  The strings are defined by the
>    authorization server.  If the value contains multiple space-delimited
>    strings, their order does not matter, and each string adds an
>    additional access range to the requested scope.

I was not aware that we follow only this guideline and don't allow scopes as "scope,scope"

So, there is nothing that needs to be fixed and sorry that I created this confusion in the first place.

**Comment by mynona** [comment]
But there is one more topic that needs to be considered with OpenID Connect.

In the AccessToken we have also "aud" which represents the clientID.

Based on the userinfo endpoint specification
https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse
a client would for example request

"email phone address profile customized1 customized2"

The OpenID Provider should then return only the valid user attributes FOR THIS CLIENT.

At the moment we are returning all user attributes because the UserManager doesn't know anything except the userID:

```
func getUser(userID: String) async throws -> VaporOAuth.OAuthUser? {
```

If we would have an optional parameter clientID the UserManager could look up the scopes for this client (AccessToken aud) and check what parameters it should return.

What do you think about extending getUser:

```
getUser(userID: String, clientID: String?)
```

I believe it would not be good if the returned attributes are automatically managed as there are also customized scopes allowed and the OpenID specification describes that out of security reasons not all requested attributes per scope must be returned. So, this should be handled by the consumer = customized UserManager

**Comment by mynona** [comment]
…and looking at our responses we are not consistent:

JWT response (example: AccessToken)

![Bildschirmfoto 2024-01-14 um 13 51 30](https://github.com/vamsii777/vapor-oauth/assets/4067383/a25ba1b2-0906-4875-97b7-c26198bc867b)

Introspection response:

```
-----------------------------
OAuthClient() introspect(accessToken:_:)
-----------------------------
Token introspection response:
Response: HTTP/1.1 200 OK
content-type: application/json; charset=utf-8
content-length: 107
set-cookie: vapor-session=n/1FXVkzTVnwy1F7PkZjpPyiTaR8MVb80blGLVEouKk=; Expires=Sun, 21 Jan 2024 12:49:59 GMT; Path=/; SameSite=Lax
connection: keep-alive
date: Sun, 14 Jan 2024 12:49:59 GMT

{"exp":1705236644,"scope":"admin openid","active":true,"username":"john_doe@something.com","client_id":"1"}
-----------------------------
```

Maybe it would make sense that we always return the scope as String with space delimeters?

scope: "admin openid email profile"

?

**Comment by mynona** [comment]
Sample JWT Tokens

AccessToken
https://auth0.com/docs/secure/tokens/access-tokens 

```
{
  "iss": "https://my-domain.auth0.com/",
  "sub": "auth0|123456",
  "aud": [
    "https://example.com/health-api",
    "https://my-domain.auth0.com/userinfo"
  ],
  "azp": "my_client_id",
  "exp": 1311281970,
  "iat": 1311280970,
  "scope": "openid profile read:patients read:admin"
}
```

Our current AccessToken:

![Bildschirmfoto 2024-01-14 um 14 03 03](https://github.com/vamsii777/vapor-oauth/assets/4067383/27b32c78-6056-4a29-9b61-fc900d2693a4)

**Comment by mynona** [comment]
Example IDToken
https://auth0.com/docs/secure/tokens/id-tokens/id-token-structure 

```
{
  "iss": "http://my-domain.auth0.com",
  "sub": "auth0|123456",
  "aud": "my_client_id",
  "exp": 1311281970,
  "iat": 1311280970,
  "name": "Jane Doe",
  "given_name": "Jane",
  "family_name": "Doe",
  "gender": "female",
  "birthdate": "0000-10-31",
  "email": "janedoe@example.com",
  "picture": "http://example.com/janedoe/me.jpg"
}
```

Our current IDToken:

![Bildschirmfoto 2024-01-14 um 14 05 36](https://github.com/vamsii777/vapor-oauth/assets/4067383/04d74f13-29b4-4911-826b-405b56d7db90)

**Comment by vamsii777** [comment]
@mynona so, about the scopes there's no need to change apart from that being string in the JWT right? The previous branch [feature/jwt](https://github.com/vamsii777/vapor-oauth/tree/feature/jwk) should be alright for the changes or [fixes/incorrect-scope-format](https://github.com/vamsii777/vapor-oauth/tree/fixes/incorrect-scope-format)?

**Comment by mynona** [comment]
In my opinion, you improved the readability of the code a lot, because you do the mapping to [String] in the ScopeValidator. This is why the issue arose in the first place because it was not obvious where "admin openid" was mapped to ["admin", "openid"]. So, I think it is an improvement.

But my second comment is more important: Extending the UserManager to return user attributes based on clientID (and therefore, return only allowed attributes for this client)

Otherwise you would return data that should not even be shared with certain clients. It's all connected.

**Comment by mynona** [comment]
After putting more thought in this: the JWT doesn't need to be changed as the payload is created by the consumer of vapor/oauth. I think, this is perfect because then every user of vapor/oauth can configure the payload themselves (and have to take care that they implement it correctly). 

This flexibility should remain on the consumer I guess.

**Comment by vamsii777** [comment]
Yes, agreed! Will check on https://github.com/vamsii777/vapor-oauth/issues/11#issuecomment-1890941857

**Comment by mynona** [comment]
Once more, sorry that I didn't see it in the first place that there happens a mapping based on spaces in the RequestObject

**Comment by vamsii777** [comment]
- Merged #15
